File: libs/rustaria-api/src/lua/log.rs
use mlua::prelude::*;
use rustaria_util::{debug, error, info, trace, warn};

use crate::lua::ctx;

pub fn register(lua: &Lua, _: ()) -> LuaResult<LuaTable<'_>> {
	lua.create_table_from([
		("trace", lua.create_function(trace)?),
		("debug", lua.create_function(debug)?),
		("info", lua.create_function(info)?),
		("warn", lua.create_function(warn)?),
		("error", lua.create_function(error)?),
	])
}

fn trace(lua: &Lua, msg: String) -> LuaResult<()> {
	trace!(target: &ctx(lua).id, "{msg}");
	Ok(())
}

fn debug(lua: &Lua, msg: String) -> LuaResult<()> {
	debug!(target: &ctx(lua).id, "{msg}");
	Ok(())
}

fn info(lua: &Lua, msg: String) -> LuaResult<()> {
	info!(target: &ctx(lua).id, "{msg}");
	Ok(())
}

fn warn(lua: &Lua, msg: String) -> LuaResult<()> {
	warn!(target: &ctx(lua).id, "{msg}");
	Ok(())
}

fn error(lua: &Lua, msg: String) -> LuaResult<()> {
	error!(target: &ctx(lua).id, "{msg}");
	Ok(())
}

File: libs/rustaria-api/src/lua/log.lua
---Logs an trace-level message
---Reserved for very low priority, often extremely verbose.
---@param msg string
---@return nil
local function trace(msg)
    -- builtin stub
end

---Logs an debug-level message.
---Reserved for lower priority information.
---@param msg string
---@return nil
local function debug(msg)
    -- builtin stub
end

---Logs an info-level message.
---Reserved for useful information.
---@param msg string
---@return nil
local function info(msg)
    -- builtin stub
end

---Logs an warning-level message.
---Reserved for hazardous situations.
---@param msg string
---@return nil
local function warn(msg)
    -- builtin stub
end

---Logs an error-level message.
---Reserved for very serious errors.
---@param msg string
---@return nil
local function error(msg)
    -- builtin
end

return { trace = trace, debug = debug, info = info, warn = warn, error = error }
